implementations/micrometer-registry-graphite/src/test/java/io/micrometer/graphite/GraphiteTagNameMapperTest.java

```diff
@@ -0,0 +1,96 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.graphite;
+
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.config.NamingConvention;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Tests for {@link GraphiteTagNameMapper}.
+ *
+ * @author Jon Schneider
+ * @author Johnny Lim
+ * @author Andrew Fitzgerald
+ */
+class GraphiteTagNameMapperTest {
+    private final GraphiteTagNameMapper nameMapper = new GraphiteTagNameMapper();
+    private final SimpleMeterRegistry registry = new SimpleMeterRegistry();
+    private final NamingConvention namingConvention = new GraphiteNamingConvention();
+
+    @Test
+    void simpleName() {
+        Meter meter = registry.counter("a.simple.counter");
+        assertThat(getName(meter)).isEqualTo("a.simple.counter");
+    }
+
+    @Test
+    void nameSubstitutions() {
+        Meter meter = registry.counter("a.name with spaces.counter");
+        assertThat(getName(meter)).isEqualTo("a.name_with_spaces.counter");
+    }
+
+    @Test
+    void invalidNameCharacters() {
+        Meter meter = registry.counter("a.very=bad;name{.counter");
+        assertThat(getName(meter)).isEqualTo("a.very_bad_name_.counter");
+    }
+
+    @Test
+    void simpleTag() {
+        Meter meter = registry.counter("a.simple.counter", "key", "value");
+        assertThat(getName(meter)).isEqualTo("a.simple.counter;key=value");
+    }
+
+    @Test
+    void multipleTags() {
+        Meter meter = registry.counter("a.simple.counter", "key", "value", "anotherKey", "another.value");
+        assertThat(getName(meter)).isEqualTo("a.simple.counter;key=value;anotherKey=another.value");
+    }
+
+    @Test
+    void emptyTagKey() {
+        Meter meter = registry.counter("a.simple.counter", "", "value");
+        assertThat(getName(meter)).isEqualTo("a.simple.counter;unspecified=value");
+    }
+
+    @Test
+    void emptyTagValue() {
+        Meter meter = registry.counter("a.simple.counter", "key", "");
+        assertThat(getName(meter)).isEqualTo("a.simple.counter;key=unspecified");
+    }
+
+    /**
+     *
+     * > Tag names must have a length >= 1 and may contain any ascii characters except ;!^=
+     * > Tag values must also have a length >= 1, and they may contain any ascii characters except ;~
+     * > UTF-8 characters may work for names and values, but they are not well tested and it is not recommended to use non-ascii characters in metric names or tags.
+     * - https://graphite.readthedocs.io/en/latest/tags.html#carbon
+     *
+     */
+    @Test
+    void terribleTagNaming() {
+        Meter meter = registry.counter("a.simple.counter", "key;!^=~ why?", "value;!^=~ why?");
+        assertThat(getName(meter)).isEqualTo("a.simple.counter;key____~ why?=value_!^=_ why?");
+    }
```

Comment:
Spaces are the biggest question mark for me, but they're not in the explicitly disallowed character list.

Reply:
Should be easy to test. JVM memory metrics ship a tag value with a space in it by default.

Reply:
Yeah, verified with a team member/contributor that spaces are valid, and it's pretty clear the source code allows it: https://github.com/graphite-project/carbon/blob/c5d7cc8a4f2eb0429594c14e035869b38cc79cf3/lib/carbon/util.py#L342-L359